### PR TITLE
render_entrypoint tag

### PR DESCRIPTION
**Pull request by alihazemfarouk**
`render_entrypoint` tag renders all chunks related to an entrypoint. This allows us to make use of Webpack 4 code splitting functionality. 

The corresponding PR for webpack-bundle-tracker (necessary for this feature to work) can be found [here](https://github.com/owais/webpack-bundle-tracker/pull/41)

### CODE REVIEW

**Comment by owais** [comment]
@alihazemfarouk Thanks for the PR. I'll review in detail but I haven't used webpack 4 yet so I'll likely have some questions. For starters, what is the difference between `render_bundle` and `render_entrypoint` ? From webpack's perspective, we used to define entrypoints in v2 and still define entrypoints in v4 so why the extra function in webpack-loader?

**Comment by alihazemfarouk** [comment]
@owais Webpack 4 introduced a feature called "Code Splitting", where essentially for every entry it traces its dependencies and optimizes the bundles it outputs for maximum efficiency (least amount of code to be shipped). The problem with `render_bundle` is that what it does it's just that, it renders a bundle, so we're unable to efficiently use the new Code Splitting webpack feature because we wouldn't know the names of the bundles webpack is going to produce, for example if react is used in entry 'a' webpack would most likely (depending on the configuration of course) generate a "node_vendors\~a.bundle.js" file containg react (and other depencies used by entry a), but then if react is used in entry 'b', webpack optimizations would likely to output 1 bundle for those common dependencies for both entry a and b and the file name will look something like this "node_vendors\~a\~b.bundle.js"
What `render_entrypoint` does is that reads the output of how webpack optimized and then produced the bundles for each entry, and renders all the bundles required by that entry. For a clearer explanation on webpack 4 code splitting please check out [this medium article](https://medium.com/dailyjs/webpack-4-splitchunks-plugin-d9fbbe091fd0)

**Comment by persocon** [comment]
👍

**Comment by greggb** [comment]
@owais Really appreciate you keeping this project alive. Is there anything you need to help move some of these PRs along? Webpack has changed dramatically over the last year and this PR would be a huge help for those of us making use of code splitting.

**Comment by alihazemfarouk** [comment]
@owais Anything I can help with to move things along for this PR?

**Comment by alihazemfarouk** [comment]
Ok so to be able to use this feature, I published this PR on pypi and npm, under the names of 'django-webpack4-loader' and 'webpack4-bundle-tracker'.
You can find the documentation for the new feature in the readme on my github fork.

@owais If at any point you review this PR and would like me to do some modifications before merging, if you think the PR is fine and ready to merge and you would like to merge it, inform me and I will delete the 2 packages I published in order not to be redundant.

**Comment by owais** [comment]
@alihazemfarouk Thanks for helping out. I've been really busy with work. Will try to reserve some time to review it next week. Sorry for the delay. Using forked packages sounds good till this is merged.

**Comment by persocon** [comment]
@owais any news on this? 🙏

**Comment by viperfx** [comment]
@alihazemfarouk are you aware of how a react router integration could work?

**Comment by alihazemfarouk** [comment]
> @alihazemfarouk thanks for your work on this. I have been waiting for this feature for ages.
> 
> Is your fork being used in your production apps? Have you been able to update the tests to verify the split functionality?
> 
> You should honestly put up some gittip or bounty, because I will happily pay for further improvements.

Thanks a lot for your nice words. Yes I have been using my fork in production. There are tests that I'd written here too that verify that the package handles loading an entrypoint correctly with all its chunks.  You can find the necessary instructions in the readme on my github fork, especially look for the parts where `EXCLUDE_RUNTIME`, `BASE_ENTRYPOINT ` and the usage of `render_entrpoint` are explained

**Comment by alihazemfarouk** [comment]
> @alihazemfarouk are you aware of how a react router integration could work? e.g. splitting based on routes defined in a react router.
> 
> ```
> export const createRoutes = (store) => ({
>   path        : '/',
>   component   : CoreLayout,
>   childRoutes : [
>     Login,
>     SignUp
>   ]
> })
> 
> export default createRoutes
> ```

Well the whole idea behind this feature is Webpack code splitting when manually specifying certain entrypoints. Ex: 
```
module.exports = {
  ...,
  entry: {
    index: "./myapp/static/src/pages/index.js",
    contact_us: "./myapp/static/src/pages/contact_us.js",
  },
  ...,
  plugins: [new BundleTracker({ filename: "./webpack-stats.json" })]
};
```
as it reads from the stats file the plugin `BundleTracker` produces. 

If you have any thoughts on how to improve this please share

**Comment by viperfx** [comment]
@alihazemfarouk I installed `django-webpack4-loader`

`ModuleNotFoundError: No module named 'webpack_loader'`

**Comment by viperfx** [comment]
@alihazemfarouk also the `examples` folder does not contain an example for entry_point. Could you add that?

**Comment by alihazemfarouk** [comment]
> @alihazemfarouk I installed `django-webpack4-loader`
> 
> However, I got the following issue
> 
> `ModuleNotFoundError: No module named 'webpack_loader'`

The package is working for me both locally and in production apps. Try uninstalling both `django-webpack4-loader` and `django-webpack-loader` and reinstall `django-webpack4-loader` again.

**Comment by alihazemfarouk** [comment]
> @alihazemfarouk also the `examples` folder does not contain an example for entry_point. Could you add that?

Sure I can. The first chance I get some time off I will do that.

**Comment by viperfx** [comment]
Quick update and potential bug I found (could be my misconfiguration) for BASE_ENTRYPOINT

My webpack produces the following entry point
```
Entrypoint main = vendors.css vendors.js main.css main.js
```

My current config is this
```
WEBPACK_LOADER = {
    'DEFAULT': {
        'CACHE': False,
        'BUNDLE_DIR_NAME': "bundles/",
        'STATS_FILE': os.path.join(BASE_DIR, 'webpack-stats.json'),
        'POLL_INTERVAL': 0.1,
        'TIMEOUT': None,
        'EXCLUDE_RUNTIME': False,
        'BASE_ENTRYPOINT': 'main'
    }
}
```

I am including my tags similar to

```
{% load render_entrypoint from webpack_loader %}

<html>
  <head>
    {% render_entrypoint 'main' 'css' %}
  </head>
  <body>
    ....
    {% render_entrypoint 'main' 'js' %}
  </body>
</head>
```

In the view source, I see two includes like this.

```
<link type="text/css" href="/static/bundles/vendors.css" rel="stylesheet" />
   <script type="text/javascript" src="/static/bundles/vendors.js" ></script>
  <link type="text/css" href="/static/bundles/main.css" rel="stylesheet" />
  <script type="text/javascript" src="/static/bundles/main.js" ></script>


```

File: README.md [hunk 1]
```diff
@@ -166,6 +168,19 @@ and your webpack config is located at `/home/src/webpack.config.js`, then the va
...
+
+`EXCLUDE_RUNTIME` is meant to be used with `render_entrypoint`. When creating multi-page applications, it's common to want to
+split a single runtime file that is used by all chunks. You can do that by using `{% render_bundle 'runtime' %}` in your base HTML file and setting `EXCLUDE_RUNTIME` to `True` in order to not include it again when using `{% render_entrypoint 'example_entry_point' %}`
+
+<br>
+
+#### BASE_ENTRYPOINT
+
+`BASE_ENTRYPOINT` is meant to be used with `render_entrypoint`. When creating multi-page applications, it's common to want to
+include common js in the base HTML. If the main entrypoint's name is `main`, you can do that by including `{% render_entrypoint 'main' %}` in your base HTML file. Now in another entrypoints (that extend the base HTML file), there might be some chunks that were already included in `main`, that means they would be included twice in the final rendered HTML, to avoid that, set `BASE_ENTRYPOINT` to `'main'`, then any duplicate chunks between an entrypoint and the main entrypoint would be included only once.
```
Comment (jpic) [inline README.md]: Duplicate "When creating multi-page applications, it's common to want to
￼	include common js in the base HTML"

File: README.md [hunk 2]
```diff
@@ -188,6 +203,8 @@ INSTALLED_APPS = (
 
 ### Templates
 
+### render_bundle
+
```
Comment (jpic) [inline README.md]: I think those need quadruple dash

File: README.md [hunk 3]
```diff
@@ -212,6 +229,46 @@ INSTALLED_APPS = (
...
+  ...,
+  entry: {
+    index: "./myapp/static/src/pages/index.js",
+    contact_us: "./myapp/static/src/pages/contact_us.js",
+  },
+  ...,
+  plugins: [new BundleTracker({ filename: "./webpack-stats.json" })]
+};
+```
+
```
Comment (jpic) [inline README.md]: Seems like this newline belonged to before the code block

**Comment by alihazemfarouk** [comment]
> Quick update and potential bug I found (could be my misconfiguration) for BASE_ENTRYPOINT
> 
> My webpack produces the following entry point
> 
> ```
> Entrypoint main = vendors.css vendors.js main.css main.js
> ```
> My current config is this
> 
> ```
> WEBPACK_LOADER = {
>     'DEFAULT': {
>         'CACHE': False,
>         'BUNDLE_DIR_NAME': "bundles/",
>         'STATS_FILE': os.path.join(BASE_DIR, 'webpack-stats.json'),
>         'POLL_INTERVAL': 0.1,
>         'TIMEOUT': None,
>         'EXCLUDE_RUNTIME': False,
>         'BASE_ENTRYPOINT': 'main'
>     }
> }
> ```
> I am including my tags similar to
> 
> ```
> {% load render_entrypoint from webpack_loader %}
> 
> <html>
>   <head>
>     {% render_entrypoint 'main' 'css' %}
>   </head>
>   <body>
>     ....
>     {% render_entrypoint 'main' 'js' %}
>   </body>
> </head>
> ```
> In the view source, I see two includes like this.
> 
> ```
> <link type="text/css" href="/static/bundles/vendors.css" rel="stylesheet" />
>    <script type="text/javascript" src="/static/bundles/vendors.js" ></script>
>   <link type="text/css" href="/static/bundles/main.css" rel="stylesheet" />
>   <script type="text/javascript" src="/static/bundles/main.js" ></script>
> ```

I don't understand the problem. What you're showing is the correct output, all the files, whether js or css, consisting that entrypoint, are rendered only once.

File: README.md (hunk 1 above)
Comment (alihazemfarouk) [reply README.md]: That's not by mistake. If you have a clearer way to explain it, please suggest it to me, or do it yourself and push it here

**Comment by viperfx** [comment]
I think you misunderstood my comment.

It gets rendered twice. The 4 script tags for both JS and CSS, gets
included in bottom and top of the page.

**Comment by alihazemfarouk** [comment]
> I think you misunderstood my comment. It gets rendered twice. The 4 script tags for both JS and CSS, gets included in bottom and top of the page.
> […](#)

Ahh I had not implemented the filtering by extension part yet, and yet I included it in the documentation. Silly me.
Thanks for making me notice it. I added the feature and added a test case verifying it, it is now part of this PR and is also published at pypi with version 0.0.5. Please upgrade django-webpack4-loader and tell me if it is still not working for you

**Comment by viperfx** [comment]
Thanks for the update. That fixed the issue.

**Comment by bhrutledge** [comment]
For folks waiting on this PR, I've rounded up [relevant links and workarounds](https://github.com/owais/webpack-bundle-tracker/pull/41#issuecomment-463709521) in a comment on PR https://github.com/owais/webpack-bundle-tracker/pull/41, on which this depends.

**Comment by danihodovic** [comment]
Since this repo is largely unmaintained is @alihazemfarouk the go-to solution now?

**Comment by benpetty** [comment]
> Since this repo is largely unmaintained is @alihazemfarouk fork the go-to solution now?

We are using that fork in production and works great 👍

**Comment by fjsj** [comment]
Closing this in favor of https://github.com/django-webpack/django-webpack-loader/pull/261
Which shares some of the same commits

**Pull request closed without merging**